utils/httpbody/httpbody.go

```diff
@@ -0,0 +1,53 @@
+package httpbody
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+// BackupRequestBody returns a copy of the Request.Body, leaving it untouched
+// so the passed Request.Body will be still readable.
+func BackupRequestBody(req *http.Request) (io.ReadCloser, error) {
+	content, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Body = ioutil.NopCloser(bytes.NewBuffer(content))
+	return ioutil.NopCloser(bytes.NewBuffer(content)), nil
+}
+
+// ReadResponseAndRestore returns the content of the passed Response.Body,
+// and restores it to allow its reuse (e.g. when read from a nested Transport)
+func ReadResponseAndRestore(resp *http.Response) ([]byte, error) {
+	reader, content, err := readAndGetNew(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.Body = reader
```

Comment:
what about closing the original `Response.Body` in the helper?
```suggestion
	resp.Body = reader
	resp.Body.Close()
```

Reply:
57eca7e5b